### ISSUE TITLE
CosmWasm v1

### ISSUE THREAD

**Issue by reuvenpo** [scrtlabs/SecretNetwork#770]
This task is pretty open ended - there are many improvements and changes made to cosmwasm upstream since we forked them at 0.10 ([changelog](https://github.com/CosmWasm/cosmwasm/blob/main/CHANGELOG.md)).

We need to fully map out the features we want/can backport in the timeframe we have, and replace this line with a list of check-boxes listing them.

**Comment by mohammedpatla** [comment]
@reuvenpo Just on the CosmWasm topic, I heard from the Juno network that 0.14 and 1.0 are very different and that 1.0 will take 2-3 months to release. Probably close to or outside our time frame for IBC.

But if we go to 0.14 we miss a ton of stables, what's the plan to use for the IBC version?

**Comment by assafmo** [comment]
@mohammedpatla we're going to update the cosmwasm-std package and leave the API/ABI on-chain as is, therefore we can always upgrade cosmwasm-std to 0.14 and then to 1.0 without needing to hard fork.

**Comment by reuvenpo** [comment]
I'll add that if we do make ABI changes, we'll provide a runtime that handles both old and new contracts.
Keep in mind that we we always have our own flavor of CosmWasm because of the changes to inter-contract callbacks and queries, as well as the disabled features. When we talk about "upgrading to version X" we mean upgrading the interfaces visible to the programmer, making them conform as much as possible to the upstream offering.

**Comment by assafmo** [comment]
For reference: https://github.com/terra-money/mainnet/wiki/Columbus-5-Contract-Migration-Guide

**Comment by assafmo** [comment]
Updated TODOs:
- [ ] In `ContractInstance`/`start_engine()`, check whether the wasm is v0.10 or v0.16 ([using this export](https://github.com/CosmWasm/cosmwasm/blob/v0.16.0/packages/vm/README.md?plain=1#L11-L28)) and store it in `ContractInstance`, then (after the crypto stuff) according to the cosmwasm version, initialize the `env` object that is given to the contract in:
  - [ ]  [`init`](https://github.com/enigmampc/SecretNetwork/blob/c05c74f9eddc6213515a39dd43c7446b08e3cea8/cosmwasm/packages/wasmi-runtime/src/wasm/contract_operations.rs#L103)
  - [ ] [`handle`](https://github.com/enigmampc/SecretNetwork/blob/c05c74f9eddc6213515a39dd43c7446b08e3cea8/cosmwasm/packages/wasmi-runtime/src/wasm/contract_operations.rs#L221)
  - [ ] and [`query`](https://github.com/enigmampc/SecretNetwork/blob/c05c74f9eddc6213515a39dd43c7446b08e3cea8/cosmwasm/packages/wasmi-runtime/src/wasm/contract_operations.rs#L300) (in v0.16 `query` has `env` as well).
- [ ] Also, invoke the right wasm function:
  - [ ] v0.10 `init` or v0.16 `instantiate` ([link](https://github.com/enigmampc/SecretNetwork/blob/c05c74f9eddc6213515a39dd43c7446b08e3cea8/cosmwasm/packages/wasmi-runtime/src/wasm/runtime/engine.rs?plain=1#L45)).
  - [ ] v0.10 `handle` or v0.16 `execute` ([link](https://github.com/enigmampc/SecretNetwork/blob/c05c74f9eddc6213515a39dd43c7446b08e3cea8/cosmwasm/packages/wasmi-runtime/src/wasm/runtime/engine.rs?plain=1#L111))
- [ ] Afterwards, translate the v0.16 output into v0.10 output so that the Go code will handle it nicley.
  - [ ] Add plaintext_log to cosmwasm-std v0.16
  - [ ] Handle v0.16 `submessages`. TODO: See how cosmwasm implemented this, we might want to piggyback off `messages` and handle it in the Go code, or maybe we can handle it right there inside the enclave by converting them to normal messages.   
- [ ] Fix wasm validation. The new v0.16 API needs [these](https://github.com/CosmWasm/cosmwasm/blob/c14937849cedf7c70f85c15b90053bdcf01e8a1c/packages/vm/src/compatibility.rs?plain=1#L12-L41) while we currently have the v0.10 API which validates [these](https://github.com/enigmampc/SecretNetwork/blob/c05c74f9eddc6213515a39dd43c7446b08e3cea8/cosmwasm/packages/sgx-vm/src/compatability.rs?plain=1#L11-L34). In order to support both in parallel, we need to fix the validation code.
- [ ] Implement new v0.16 API
  - [x] `db_read` - we already have this from v0.10
  - [x] `db_write` - we already have this from v0.10
  - [x] `db_remove` - we already have this from v0.10
  - [ ] `addr_validate`
  - [ ] `addr_canonicalize` - need to alias from `canonicalize_address` (v0.10)
  - [ ] `addr_humanize` - need to alias from `humanize_address` (v0.10)
  - [ ] `secp256k1_verify`
  - [ ] `secp256k1_recover_pubkey`
  - [ ] `ed25519_verify`
  - [ ] `ed25519_batch_verify`
  - [ ] `debug` - debug prints from inside the contract
  - [x] `query_chain` - we already have this from v0.10

**Comment by assafmo** [comment]
Closing in favor of https://github.com/orgs/scrtlabs/projects/3 and https://github.com/scrtlabs/SecretNetwork/pull/930
